contrib/lib-autopilot.py

```diff
@@ -0,0 +1,289 @@
+        
+        if k < 3:
+            k = 3
+        
+        self.__logger.info("IMPROOVE_NETWORK: Try to seek {} nodes which are currently bad connected".format(k))
+        candidates = heapq.nlargest(3*k, self.__generate_all_shortest_paths(), key=itemgetter(0))
+
+        for _, path in candidates:
+            result.add(path[0])
+            result.add(path[-1])
+
+        self.__logger.info("IMPROOVE_NETWORK: Found {} candidates which are currently bad connected".format(len(result)))        
+        if len(result) <= k:
+            return list(result)
+        self.__logger.info("IMPROOVE_NETWORK: Sample {} items from the candidates as was requested".format(k))
+        tmp = list(result)
+        random.shuffle(tmp)
+        return tmp[0:k]
+        
+ 
+    def __generate_all_shortest_paths(self,cutoff = 10):
+        """ 
+        this is basically a wrapper for nx.all_pair_shortest_path to have a more convenient output format
+        
+        generates all shortest paths and yields each item, the format of the output is
+        (len, [src, n1,...,n_{m-2}, dest])
+        """
+        if cutoff < 1:
+            cutoff = 10
+            self.__logger.info("cutoff value must be a positive integer. Set back to default value: 10")
+
+        all_pair_shortest_paths = nx.all_pairs_shortest_path(self.G, cutoff=cutoff)
+        for item in all_pair_shortest_paths:
+            from_node = item[0]
+            paths = item[1]
+            for destination,path in paths.items():
+                yield (len(path),path)
+                    
+    
+    def __display_nodes(self,nodes):
+        for node in nodes:
+            if "alias" in self.G.node[node]:
+                print(self.G.node[node]["alias"])
+            else:
+                print(node)
+    def get_candidates(self,k=10):
+        sub_k = math.ceil(k/4)
+        self.__logger.info("GENERATE CANDIDATES: Try to generate up to {} nodes with 4 strategies: (random, central, network Improvement, liquidity)")
+        candidats = set()
+        res = self.__generate_random_nodes(sub_k)
+        candidats=candidats.union(set(res))
+        res = self.__generate_central_nodes(sub_k)
+        candidats=candidats.union(set(res))
+        res = self.__generate_rich_nodes(sub_k)
+        candidats=candidats.union(set(res))
+        res = self.__generate_connicticity_increasing_canidates(sub_k)
+        candidats=candidats.union(set(res))
+
+        if len(candidats) > k:
+            tmp = list(candidats)
+            random.shuffle(tmp)
+            candidats = tmp[:k]
+        
+        self.__logger.info("GENERATE CANDIDATES: Found {} nodes with which channel creation is suggested".format(len(candidats)))
+        return candidats
+            
+    
+    def run(self):
+        self.__logger.info("running the autopilot on a graph with {} nodes and {} edges.".format(len(self.G.nodes()), len(self.G.edges())))
+        candidates = self.get_candidates(k=21)
+        time.sleep(1)
```

Comment:
What is this `sleep` for? It doesn't seem like we need it at all.

Reply:
I used logger and print and they got mixed so I decided while debugging I could wait a second in order to have nicer output. forgot to delete the line